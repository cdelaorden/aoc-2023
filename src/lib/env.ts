function env(key:string, defaultValue?:string):string|undefined {
  return process.env[key] ?? defaultValue ?? undefined
}

env.int = (key:string, defaultValue = 0):number => {
  return process.env[key] != null ? parseInt(process.env[key] as string, 10): defaultValue
}

env.bool = (key:string, defaultValue?:boolean):boolean => {
  const value = process.env[key]
  if (value == null) return defaultValue ?? false
  // parse strings true/TRUE, false/FALSE
  if (value.toString().toLocaleLowerCase() === 'true') return true
  if (value.toString().toLocaleLowerCase() === 'false') return false
  // parse numbers
  return Boolean(env.int(key, Number(defaultValue)))
}

env.array = (key:string, defaultValue: string[] = [], separator = ','): string[] => {
  const value = process.env[key]
  if (value == null) return defaultValue
  return value.split(separator).map(x => x.trim())
}

export default env