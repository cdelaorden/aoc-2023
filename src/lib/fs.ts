import { readFile } from 'node:fs/promises'
async function readTextFile(path: string) {
  return readFile(path, {
    encoding: 'utf-8',
  }).catch(() => '')
}

export function readSample(day: number) {
  return readTextFile(`${__dirname}/../fixtures/${day}/sample.txt`)
}

export function readInput(day: number) {
  return readTextFile(`${__dirname}/../fixtures/${day}/input.txt`)
}
