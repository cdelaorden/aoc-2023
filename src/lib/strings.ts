export const reverse = (s: string) => s.split('').reverse().join('')

export const replace =
  (search: string | RegExp, forThat: string) => (input: string) =>
    input.replaceAll(search, forThat)
