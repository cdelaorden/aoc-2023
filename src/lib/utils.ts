import { identity } from 'remeda'

export type Coord = {
  x: number
  y: number
}

export const borderAround = (c: Coord) => {
  const result: Coord[] = []
  for (let x = c.x - 1; x < c.x + 2; x++) {
    for (let y = c.y - 1; y < c.y + 2; y++) {
      if (x >= 0 && y >= 0 && (x != c.x || y != c.y)) {
        result.push({ x, y })
      }
    }
  }
  return result
}

export const sum = (xs: number[]) => xs.reduce((acc, x) => acc + x, 0)

export const compose = <T>(...funcs: Array<(x: T) => T>) => {
  return funcs.reduce((f, g) => (x: T) => g(f(x)), identity)
}

export const explode = (a = '', onChar = '') => a.split(onChar)
export const splitLines = (a = '') => a.split('\n')

export const isNumberString = (s: string) => /^\d+$/.test(s)

export function transpose<T>(matrix: T[][]) {
  if (!matrix[0]) return matrix
  // @ts-expect-error whatever
  return matrix[0].map((_col, c) => matrix.map((_row, r) => matrix[r][c]))
}

export function combine<T>(list: T[]) {
  return list.flatMap((item, i) =>
    list.slice(i + 1).map((other) => [item, other])
  )
}
