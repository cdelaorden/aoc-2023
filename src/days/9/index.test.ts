import { explode } from '@/lib/utils'
import { calculateDifferences, expandWithDifferences } from '.'

describe('calculateDifferences', () => {
  it('creates a new Array with differences', () => {
    const result = calculateDifferences([0, 3, 6, 9, 12, 15])
    expect(result).toEqual([3, 3, 3, 3, 3])
    const result2 = calculateDifferences(result)
    expect(result2).toEqual([0, 0, 0, 0])
  })
})

describe('expandWithDifferences', () => {
  it('creates an Array of dimishing differences from original row', () => {
    const result = expandWithDifferences([0, 3, 6, 9, 12, 15])
    expect(result).toEqual([
      [0, 3, 6, 9, 12, 15],
      [3, 3, 3, 3, 3],
      [0, 0, 0, 0],
    ])
  })

  it('handles neg values', () => {
    const result = expandWithDifferences(
      explode(
        '-9 -13 -17 -21 -25 -29 -33 -37 -41 -45 -49 -53 -57 -61 -65 -69 -73 -77 -81 -85 -89',
        ' '
      ).map(Number)
    )
    expect(result).toMatchSnapshot()
  })
})
