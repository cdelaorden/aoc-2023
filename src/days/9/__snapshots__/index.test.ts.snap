// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expandWithDifferences handles neg values 1`] = `
[
  [
    -9,
    -13,
    -17,
    -21,
    -25,
    -29,
    -33,
    -37,
    -41,
    -45,
    -49,
    -53,
    -57,
    -61,
    -65,
    -69,
    -73,
    -77,
    -81,
    -85,
    -89,
  ],
  [
    -4,
    -4,
    -4,
    -4,
    -4,
    -4,
    -4,
    -4,
    -4,
    -4,
    -4,
    -4,
    -4,
    -4,
    -4,
    -4,
    -4,
    -4,
    -4,
    -4,
  ],
  [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
]
`;
